File: leetcode3/변지협/v3/1665. Minimum Initial Energy to Finish Tasks.py
'''
1. 아이디어:
actual이 최소이고, minimum이 최대인 것일 수록 먼저 수행해야한다.
2. 시간복잡도:
o(nlogn)
3. 알고리즘:
정렬 후, 순회하면서 energy를 계산한다.
'''
class Solution:
    def minimumEffort(self, tasks: list[list[int]]) -> int:
        tasks = sorted(tasks, key=lambda x: x[1] - x[0])
        print(tasks)
        
        energy = 0
        for actual, minimum in tasks:
            energy = max(minimum, actual + energy)
        
        return energy

File: leetcode3/이진희/1665. Minimum Initial Energy to Finish Tasks.java
/* 

1. 아이디어 : 실제 소모되는 에너지, 일을 시작하기위해 필요한 에너지가 각각 주어졌으므로, 최소 소모되는 에너지량을 구한다.
            처음에는 1. minimum 오름차순 -> 2. actual 오름차순으로 정렬했지만, 최소 비용을 구하기 위해서는 minimum-actual 값 기준 
            오름차순으로 배열을 정렬해야함 

            남는 에너지가 많은 일부터 처리할수록 다음 task시 에너지 보충이 적어지기 때문


2. 시간복잡도 : O(NlogN)

3. 자료구조/알고리즘 : 그리디, 커스텀 정렬

 */
class Solution {
    public int minimumEffort(int[][] tasks) {
        // 순서 상관없이 모든 일을 끝낼 수 있는 "최소"에너지

        int ans = 0;
        int currEnergy = 0;

        Arrays.sort(tasks, (a,b) -> {
            return Integer.compare(b[1]-b[0], a[1]-a[0]);
        });

        for(int i=0; i<tasks.length; i++) {
            int a = tasks[i][0];
            int m = tasks[i][1];

            int diff = currEnergy-m;

            if(diff<0) {
                ans-=diff;
                currEnergy=m;
            }

            currEnergy-=a;
        }

        return ans;
    }
}